File: Project_Codes/Python Version/hypothesis_testing.py
from scipy.stats import kruskal

# ---------------------------------------
# Kruskal-Wallis Test
# ---------------------------------------
def perform_kruskal_wallis(dataframe, group_col, target_col, alpha=0.05):
    """Performs the Kruskal-Wallist test for Hypothesis Testing"""

    print(f"--- Kruskal-Wallis Test: {group_col} vs {target_col} ---")

    # Group the target variable by the grouping column
    # It creates a list of arrays for each category
    groups = [group[target_col].values for name, group in dataframe.groupby(group_col)]

    # 2. Perform the test  ('*' to unpack the list of arrays)
    stat, p_value = kruskal(*groups)

    # Print formatted results
    print(f"H-statistic: {stat:.2f}")
    print(f"P-value: {p_value:.4f}")

    # Interpret the results
    if p_value < alpha:
        print(f"Conclusion: Reject H0. '{group_col}' systematically impacts '{target_col}'.\n")
    else:
        print(f"Conclusion: Fail to reject H0. No significant impact detected.\n")

    return stat, p_value

File: Project_Codes/R Version/hypothesis_testing.R
library(dplyr)

# ---------------------------------------
# Kruskal-Wallis Test
# ---------------------------------------
perform_kruskal_wallis <- function(dataframe, group_col, target_col, alpha=0.05) {
  # Performs the Kruskal-Wallist test for Hypothesis Testing

  cat(paste0("--- Kruskal-Wallis Test: ", group_col, " vs ", target_col, " ---\n"))

  # Group the target variable by the grouping column
  # It creates a list of arrays for each category
  groups <- split(dataframe[[target_col]], dataframe[[group_col]])

  # 2. Perform the test  ('*' to unpack the list of arrays)
  test_result <- kruskal.test(groups)

  stat <- test_result$statistic
  p_value <- test_result$p.value

  # Print formatted results
  cat(sprintf("H-statistic: %.2f\n", stat))
  cat(sprintf("P-value: %.4f\n", p_value))

  # Interpret the results
  if (p_value < alpha) {
    cat(paste0("Conclusion: Reject H0. '", group_col, "' systematically impacts '", target_col, "'.\n\n"))
  } else {
    cat("Conclusion: Fail to reject H0. No significant impact detected.\n\n")
  }

  return(list(stat=stat, p_value=p_value))
}
